apps/api/src/app/events/usecases/map-trigger-recipients/map-trigger-recipients.use-case.ts

```diff
@@ -0,0 +1,139 @@
+import { Injectable } from '@nestjs/common';
+import {
+  ISubscribersDefine,
+  ITopic,
+  TriggerRecipient,
+  TriggerRecipientSubscriber,
+  TriggerRecipientTopics,
+  TriggerRecipients,
+  TriggerRecipientsPayload,
+} from '@novu/node';
+import {
+  EnvironmentId,
+  LogCodeEnum,
+  LogStatusEnum,
+  OrganizationId,
+  TopicId,
+  TriggerRecipientsTypeEnum,
+  UserId,
+} from '@novu/shared';
+
+import { MapTriggerRecipientsCommand } from './map-trigger-recipients.command';
+
+import { CreateLog, CreateLogCommand } from '../../../logs/usecases/create-log';
+import { GetTopicSubscribersCommand, GetTopicSubscribersUseCase } from '../../../topics/use-cases';
+
+interface ILogTopicSubscribersPayload {
+  environmentId: EnvironmentId;
+  organizationId: OrganizationId;
+  topicId: TopicId;
+  transactionId: string;
+  userId: UserId;
+}
+
+@Injectable()
+export class MapTriggerRecipients {
+  constructor(private createLog: CreateLog, private getTopicSubscribers: GetTopicSubscribersUseCase) {}
+
+  async execute(command: MapTriggerRecipientsCommand): Promise<ISubscribersDefine[]> {
+    const { environmentId, organizationId, recipients, transactionId, userId } = command;
+
+    const mappedRecipients = Array.isArray(recipients) ? recipients : [recipients];
+
+    const simpleSubscribers: ISubscribersDefine[] = this.findSubscribers(mappedRecipients);
+
+    const topicSubscribers: ISubscribersDefine[] = await this.getSubscribersFromAllTopics(
+      transactionId,
+      environmentId,
+      organizationId,
+      userId,
+      mappedRecipients
+    );
+
+    return [...simpleSubscribers, ...topicSubscribers];
+  }
+
+  private async getSubscribersFromAllTopics(
+    transactionId: string,
+    environmentId: EnvironmentId,
+    organizationId: OrganizationId,
+    userId: UserId,
+    recipients: TriggerRecipients
+  ): Promise<ISubscribersDefine[]> {
+    const topics = this.findTopics(recipients);
+
+    const topicSubscribers: ISubscribersDefine[] = [];
+
+    for (const topic of topics) {
+      try {
+        const getTopicSubscribersCommand = GetTopicSubscribersCommand.create({
+          environmentId,
+          topicId: topic?.topicId || '',
+          organizationId,
+          userId,
+        });
+        const response = await this.getTopicSubscribers.execute(getTopicSubscribersCommand);
```

Comment:
I wonder how this will work with large topics, was thinking maybe we can use mongo db `cursor` functionality to iterate over larger collections? I think David implemented something similar on the `broadcast` endpoint

Reply:
Topic size limitation is yet to be implemented also, so we can take that in consideration. But yeah, large topics scenario hasn't yet been addressed.